pkg/attachutils/simpleattach.go

```diff
@@ -0,0 +1,385 @@
+	)
+
+	if clientStdin != nil {
+		wg.Add(1)
+	}
+
+	if clientStdout != nil {
+		wg.Add(1)
+	}
+
+	if clientStderr != nil {
+		wg.Add(1)
+	}
+
+	// Connect stdin of container to the http conn.
+	go func() {
+		if clientStdin == nil {
+			return
+		}
+		log.Debugf("attach: stdin: begin")
+
+		var err error
+		if tty {
+			_, err = copyEscapable(svrStdin, clientStdin, keys)
+		} else {
+			_, err = io.Copy(svrStdin, clientStdin)
+		}
+		if err == io.ErrClosedPipe {
+			err = nil
+		}
+		if err != nil {
+			log.Printf("attach: stdin: %s\n", err)
+			errors <- err
+		}
+		if !tty {
+			svrStdin.Close()
+		}
+		log.Debugf("attach: stdin: end")
+		wg.Done()
+	}()
+
+	attachStream := func(name string, clientWriter io.Writer, svrReader io.Reader) {
+		log.Debugf("attach: %s: begin", name)
+		_, err := io.Copy(clientWriter, svrReader)
+		if err == io.ErrClosedPipe {
+			err = nil
+		}
+		if err != nil {
+			log.Errorf("attach: %s: %v", name, err)
+			errors <- err
+		}
+		// Make sure stdin gets closed
+		if clientStdin != nil {
+			clientStdin.Close()
+		}
+		log.Debugf("attach: %s: end", name)
+		wg.Done()
+	}
+
+	go attachStream("stdout", clientStdout, svrStdout)
+	go attachStream("stderr", clientStderr, svrStderr)
+
+	done := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+	<-done
+	close(errors)
+	for err := range errors {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Code c/c from io.Copy() modified to handle escape sequence
+func copyEscapable(dst io.Writer, src io.ReadCloser, keys []byte) (written int64, err error) {
+	if len(keys) == 0 {
+		// Default keys : ctrl-p ctrl-q
+		keys = []byte{16, 17}
+	}
+	buf := make([]byte, 32*1024)
+	for {
+		nr, er := src.Read(buf)
+		if nr > 0 {
+			// ---- Docker addition
+			for i, key := range keys {
+				if nr != 1 || buf[0] != key {
+					break
+				}
+				if i == len(keys)-1 {
+					if err := src.Close(); err != nil {
+						return 0, err
+					}
+					return 0, nil
+				}
+				nr, er = src.Read(buf)
+			}
+			// ---- End of docker
+			nw, ew := dst.Write(buf[0:nr])
+			if nw > 0 {
+				written += int64(nw)
+			}
+			if ew != nil {
+				err = ew
+				break
+			}
+			if nr != nw {
+				err = io.ErrShortWrite
+				break
+			}
+		}
+		if er == io.EOF {
+			break
+		}
+		if er != nil {
+			err = er
+			break
+		}
+	}
+	return written, err
+}
+
+// HijackConnection is a helper function to call from HTTP middleware that has
+// access to the http response writer.
+func HijackConnection(rw http.ResponseWriter, r *http.Request) HijackClosure {
+	_, upgrade := r.Header["Upgrade"]
+
+	hijacker, ok := rw.(http.Hijacker)
+	if !ok {
+		return nil
+	}
+
+	conn, _, err := hijacker.Hijack()
+	if err != nil {
+		return nil
+	}
+
+	// GetStreams hijacks the the connection from the http handler. This code originated
+	// in the docker engine-api image router.
+	setupStreams := func() (io.ReadCloser, io.Writer, io.Writer, error) {
+		// set raw mode
+		conn.Write([]byte{})
```

Comment:
Please clarify.  What is this setting and how?  It's writing with length 0; which we believe to be a NOOP.